workflow/controller/pod_cleanup.go

```diff
@@ -1,22 +1,142 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/sirupsen/logrus"
 
 	"github.com/argoproj/argo-workflows/v3/workflow/common"
 
 	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	"github.com/argoproj/argo-workflows/v3/workflow/controller/indexes"
 )
 
-func (woc *wfOperationCtx) queuePodsForCleanup() {
+func (woc *wfOperationCtx) isImmediateSkipQueues() bool {
+	return woc.rateLimiter != nil && woc.orig.Spec.PodGC != nil && woc.orig.Spec.PodGC.Strategy == wfv1.PodGCOnPodCompletion
+}
+
+// @podGC is used to generate the label selector @labelPodCompleted is used to determine if
+// we should add a common.LabelKeyCompleted=true label
+func getLabelSelector(wfName string, podGC *wfv1.PodGC, labelCompleted bool) metav1.LabelSelector {
+	// upper limit on memory usage set
+	labels := make(map[string]string)
+	labels[common.LabelKeyWorkflow] = wfName
+	if labelCompleted {
+		labels[common.LabelKeyCompleted] = "true"
+	}
+
+	expressions := []metav1.LabelSelectorRequirement{}
+
+	if podGC != nil && podGC.LabelSelector != nil {
+		for key, value := range podGC.LabelSelector.MatchLabels {
+			labels[key] = value
+		}
+		expressions = append(expressions, podGC.LabelSelector.MatchExpressions...)
+	}
+	return metav1.LabelSelector{MatchLabels: labels, MatchExpressions: expressions}
+}
+
+func (woc *wfOperationCtx) runImmediateCleanup(pod *apiv1.Pod, podGC *wfv1.PodGC, workflowPhase wfv1.WorkflowPhase) error {
+	/// we have finished the cleanup previously
+	if woc.finishedCleanup {
+		return nil
+	}
+
+	log := woc.log.WithField("podName", pod.Name)
+	log.Infof("Running PodGC strategy PodGCOnWorkflowCompletion for %s", pod.Name)
+	pods := woc.controller.kubeclientset.CoreV1().Pods(pod.Namespace)
+	wfName := woc.execWf.ObjectMeta.Name
+	podGCSelector, err := podGC.GetLabelSelector()
+	if err != nil {
+		return err
+	}
+	woc.rateLimiter.Wait()
+	// check if we need to label the pod as completed first and do so if we must
+	if determinePodCleanupAction(podGCSelector, pod.Labels, wfv1.PodGCOnPodCompletion, workflowPhase, pod.Status.Phase) == labelPodCompleted {
+		_, err := pods.Patch(
+			context.Background(),
+			pod.Name,
+			types.MergePatchType,
+			[]byte(`{"metadata": {"labels": {"workflows.argoproj.io/completed": "true"}}}`),
+			metav1.PatchOptions{},
+		)
+		if err != nil {
+			log.Errorf("was not able to patch pods with completed as true for pods %s in workflow %s due to %s hence abandoning deletion", pod.Name, woc.execWf.ObjectMeta.Name, err)
+			return err
+		}
+	}
+
+	var lselector metav1.LabelSelector
+	if woc.wf.Status.Fulfilled() {
+		log.Info("workflow is fulfilled, deleting everything which may be deleted as per PodGC.Strategy.LabelSelector")
+		lselector = getLabelSelector(wfName, podGC, false)
+	} else {
+		log.Info("deleting existing pods marked as completed")
+		lselector = getLabelSelector(wfName, podGC, true)
+	}
```

Comment:
Would it make sens to not have the else statement but instead have the else as default:
```suggestion
	lselector := getLabelSelector(wfName, podGC, true)
	if woc.wf.Status.Fulfilled() {
		log.Info("workflow is fulfilled, deleting everything which may be deleted as per PodGC.Strategy.LabelSelector")
		lselector = getLabelSelector(wfName, podGC, false)
	}
```

Reply:
Yeah we can do that, easy.